lightning-block-sync/src/http_clients.rs

```diff
@@ -0,0 +1,1080 @@
+}
+
+impl HttpClient {
+	/// Opens a connection to an HTTP endpoint.
+	fn connect<E: ToSocketAddrs>(endpoint: E) -> std::io::Result<Self> {
+		let address = match endpoint.to_socket_addrs()?.next() {
+			None => {
+				return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "could not resolve to any addresses"));
+			},
+			Some(address) => address,
+		};
+		let stream = std::net::TcpStream::connect_timeout(&address, Duration::from_secs(1))?;
+		stream.set_read_timeout(Some(Duration::from_secs(2)))?;
+		stream.set_write_timeout(Some(Duration::from_secs(1)))?;
+
+		#[cfg(feature = "tokio")]
+		let stream = TcpStream::from_std(stream)?;
+
+		Ok(Self { stream })
+	}
+
+	/// Sends a `GET` request for a resource identified by `uri` at the `host`.
+	async fn get<F>(&mut self, uri: &str, host: &str) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		write!(self.stream,
+			"GET {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
+			 Connection: keep-alive\r\n\
+			 \r\n", uri, host)?;
+
+		let bytes = read_http_resp(&self.stream, MAX_HTTP_RESPONSE_LEN).await?;
+		F::try_from(bytes)
+	}
+
+	/// Sends a `POST` request for a resource identified by `uri` at the `host` using the given HTTP
+	/// authentication credentials.
+	///
+	/// The request body consists of the provided JSON `content`. Returns the response body in `F`
+	/// format.
+	async fn post<F>(&mut self, uri: &str, host: &str, auth: &str, content: serde_json::Value) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		let content = content.to_string();
+		write!(self.stream,
+			"POST {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
+			 Authorization: {}\r\n\
+			 Connection: keep-alive\r\n\
+			 Content-Type: application/json\r\n\
+			 Content-Length: {}\r\n\
+			 \r\n\
+			 {}", uri, host, auth, content.len(), content)?;
+
+		let bytes = read_http_resp(&self.stream, MAX_HTTP_RESPONSE_LEN).await?;
+		F::try_from(bytes)
+	}
+}
+
+async fn read_http_resp(mut socket: &TcpStream, max_resp: usize) -> std::io::Result<Vec<u8>> {
+	let mut resp = Vec::new();
+	let mut bytes_read = 0;
+	macro_rules! read_socket { () => { {
+		#[cfg(feature = "tokio")]
+		let bytes_read = socket.read(&mut resp[bytes_read..]).await?;
+		#[cfg(not(feature = "tokio"))]
+		let bytes_read = socket.read(&mut resp[bytes_read..])?;
+		if bytes_read == 0 {
+			return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "zero bytes read"));
+		} else {
+			bytes_read
+		}
+	} } }
+
+	let mut actual_len = 0;
+	let mut ok_found = false;
+	let mut chunked = false;
+	// We expect the HTTP headers to fit in 8KB, and use resp as a temporary buffer for headers
+	// until we know our real length.
+	resp.extend_from_slice(&[0; 8192]);
+	'read_headers: loop {
+		if bytes_read >= 8192 {
+			return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "headers too large"));
+		}
+		bytes_read += read_socket!();
+		for line in resp[..bytes_read].split(|c| *c == '\n' as u8 || *c == '\r' as u8) {
```

Comment:
It seems to me that this inner loop looks at all lines in the response buffer, even if they are beyond the end of the headers.  So if there's something that looks like a header near the top of the body, it will get confused.